component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/service/LocalCacheService.java

```diff
@@ -17,71 +17,236 @@
 
 import java.io.ObjectStreamException;
 import java.io.Serializable;
+import java.util.List;
+import java.util.Map.Entry;
 import java.util.Objects;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.function.Predicate;
 import java.util.function.Supplier;
 
+import org.talend.sdk.component.api.configuration.Option;
 import org.talend.sdk.component.api.service.cache.LocalCache;
+import org.talend.sdk.component.api.service.configuration.Configuration;
 import org.talend.sdk.component.runtime.serialization.SerializableService;
 
-import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
 
-@AllArgsConstructor
-public class LocalCacheService implements LocalCache, Serializable {
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static java.util.stream.Collectors.toList;
 
+
+/**
+ * Implementation of LocalCache with in memory conccurent map.
+ * @param <T>   the type of cached data.
+ */
+@RequiredArgsConstructor
+public class LocalCacheService<T> implements LocalCache<T>, Serializable {
+
+    /** plugin name for this cache */
     private final String plugin;
 
-    private final ConcurrentMap<String, Element> cache = new ConcurrentHashMap<>();
 
+    private final ConcurrentMap<String, ElementImpl<T>> cache = new ConcurrentHashMap<>();
+
+    @Configuration("talend.component.manager.services.cache.eviction")
+    private Supplier<CacheConfiguration> configuration;
+
+    // scheduler we use to evict tokens
+    private volatile ScheduledExecutorService threadService;
+
+    /**
+     * Evict an object.
+     * @param key key to evict.
+     */
     @Override
     public void evict(final String key) {
-        cache.remove(internalKey(key));
+        final ElementImpl<T> removedElement = cache.remove(internalKey(key));
+        if (removedElement != null) {
+            removedElement.release();
+        }
     }
 
     @Override
-    public <T> void evictIfValue(final String key, final T expected) {
-        cache.remove(internalKey(key), new Element(expected, 0));
+    public void evictIfValue(final String key, final T expected) {
+        final String realKey = internalKey(key);
+
+        // use compute to be able to call release.
+        cache.compute(realKey,
+                (String oldKey , ElementImpl<T> oldElement) -> {
+                    if (oldElement != null && Objects.equals(oldElement.getValue(), expected)) {
```

Comment:
I suspect it can be neat to benefit the fact you have oldElement to check also the timestamp if not null to do a normal eviction

Reply:
??? in this function, we force the eviction whatever timestamp value.

Reply:
if it is equals, when it is not equals (so the cached value is not the expected one for that key) the proposal was to check the timestamp to do the eviction without waiting for the background thread (which can be inactive in this env)

Reply:
ok, get it !